x-pack/legacy/plugins/apm/public/utils/formatters/duration.ts

```diff
@@ -0,0 +1,161 @@
+      asDecimal(moment.duration(value / 1000).asMinutes())
+  },
+  seconds: {
+    label: i18n.translate('xpack.apm.formatters.secondsTimeUnitLabel', {
+      defaultMessage: 's'
+    }),
+    convert: (value: number) =>
+      asDecimal(moment.duration(value / 1000).asSeconds())
+  },
+  milliseconds: {
+    label: i18n.translate('xpack.apm.formatters.millisTimeUnitLabel', {
+      defaultMessage: 'ms'
+    }),
+    convert: (value: number) =>
+      asInteger(moment.duration(value / 1000).asMilliseconds())
+  },
+  microseconds: {
+    label: i18n.translate('xpack.apm.formatters.microsTimeUnitLabel', {
+      defaultMessage: 'μs'
+    }),
+    convert: (value: number) => asInteger(value)
+  }
+};
+
+function convertTo(
+  timeUnit: string,
+  value: Maybe<number>,
+  withUnit: boolean,
+  defaultValue: string
+) {
+  const unit = durationUnit[timeUnit];
+  if (!unit || value == null) {
+    return defaultValue;
+  }
+
+  const message = SPACE + unit.label;
+
+  const convertedValue = unit.convert(value);
+  return `${convertedValue}${withUnit ? message : ''}`;
+}
+
+export const asHours = (
+  value: Maybe<number>,
+  { withUnit = true, defaultValue = NOT_AVAILABLE_LABEL }: FormatterOptions = {}
+) => convertTo('hours', value, withUnit, defaultValue);
+
+export const asMinutes = (
+  value: Maybe<number>,
+  { withUnit = true, defaultValue = NOT_AVAILABLE_LABEL }: FormatterOptions = {}
+) => convertTo('minutes', value, withUnit, defaultValue);
+
+export const asSeconds = (
+  value: Maybe<number>,
+  { withUnit = true, defaultValue = NOT_AVAILABLE_LABEL }: FormatterOptions = {}
+) => convertTo('seconds', value, withUnit, defaultValue);
+
+export const asMillis = (
+  value: Maybe<number>,
+  { withUnit = true, defaultValue = NOT_AVAILABLE_LABEL }: FormatterOptions = {}
+) => convertTo('milliseconds', value, withUnit, defaultValue);
+
+export const asMicros = (
+  value: Maybe<number>,
+  { withUnit = true, defaultValue = NOT_AVAILABLE_LABEL }: FormatterOptions = {}
+) => convertTo('microseconds', value, withUnit, defaultValue);
+
+export type TimeFormatter = (
+  value: Maybe<number>,
+  options?: FormatterOptions
+) => string;
+
+type TimeFormatterBuilder = (max: number) => TimeFormatter;
+
+export const getDurationFormatter: TimeFormatterBuilder = memoize(
+  (max: number) => {
+    const unit = getDurationUnit(max);
+    switch (unit) {
+      case 'h':
+        return asHours;
+      case 'm':
+        return asMinutes;
+      case 's':
+        return asSeconds;
+      case 'ms':
+        return asMillis;
+      case 'us':
+        return asMicros;
+    }
+  }
+);
+
+export function getDurationUnit(max: number) {
+  const toMicroseconds = (timeUnit: TimeUnit, multiplier: number = 1000) =>
+    moment.duration(1, timeUnit).asMilliseconds() * multiplier;
+
+  if (max > toMicroseconds('hours')) {
+    return 'h';
+  }
+  if (max > toMicroseconds('minutes')) {
+    return 'm';
+  }
+  if (max > toMicroseconds('seconds', 10000)) {
```

Comment:
This looks wrong. According to this 1 second is 10000000 microseconds. But it should be 1000000 microseconds?

Reply:
These are the original cut-offs:
```
const HOURS_CUT_OFF = 3600000000;
const MINUTES_CUT_OFF = 60000000;
const SECONDS_CUT_OFF = 10 * 1000000; // = 10000000
const MILLISECONDS_CUT_OFF = 10 * 1000; // = 10000
```

1 second in milliseconds is `1000` to have to correct value in microseconds I have to multiply it by `10000` to get the same `10000000` as the cut-off.